### PR TITLE
Added feature to flatten ndjson

### PR DESCRIPTION
**Pull request by utkarsharma2**
closes: #73

Provided users with options to flatten NDJSON file when the values are nested. 

Options to flatten NDJson:

- provide path to the attribute 
- Use lambda function to create a custom logic. Here Lambda function have access to complete row as it first param.

Example File:
```

{
 "name": "dave",
 "address" : {
    "city" : "landon",
    "country": "UK",
    "street": "221B Bakers Street"
  }
}
{
 "name": "jane",
 "address" : {
    "city" : "Landon",
    "country": "UK",
    "street": "221B Bakers Street" 
  }
}

```
**API**
```
load_file(
   "path/to/file/somefile.ndjson",
   "output_table": <astro table object>,
   "flatten_ndjson": {
      "address":  "address.street",
      "name": lambda x: x.toUpper()
   }
)
```

**Results:**

Name   Address
DAVE    221B Bakers Street
JANE    221B Bakers Street

### CODE REVIEW

File: src/astro/sql/operators/agnostic_load_file.py [hunk 1]
```diff
@@ -132,16 +135,66 @@ def _load_dataframe(self, path):
...
+                return deserialiser[file_type](
+                    stream, **deserialiser_params.get(file_type, {})
+                )
+
+    def process_ndjson(self, rows):
+        results = {}
+        for row in rows:
+
+            row = json.loads(row)
+            if len(row) == 0 or len(row.keys()) != 8:
```
Comment (utkarsharma2) [inline src/astro/sql/operators/agnostic_load_file.py:162]: len(row.keys()) != 8:

Comment (utkarsharma2) [inline src/astro/sql/operators/agnostic_load_file.py:162]: TO DO: Remove line 162, was a work around for the testcases to pass. We need to update a data in testcases.

File: src/astro/sql/operators/agnostic_load_file.py [hunk 2]
```diff
@@ -132,16 +135,66 @@ def _load_dataframe(self, path):
...
+                        df = pd.DataFrame(self.process_ndjson(rows))
+                    else:
+                        df = df.append(pd.DataFrame(self.process_ndjson(rows)))
+                    print("count : ", count)
+                    count += 1
+                return df
+            else:
+                return deserialiser[file_type](
+                    stream, **deserialiser_params.get(file_type, {})
+                )
```
Comment (dimberman) [inline src/astro/sql/operators/agnostic_load_file.py:155]: Can this all be a helper function?

File: src/astro/sql/operators/agnostic_load_file.py [hunk 3]
```diff
@@ -132,16 +135,66 @@ def _load_dataframe(self, path):
...
+                while True:
+                    rows = stream.readlines(DEFAULT_CHUNK_SIZE)
+                    if len(rows) == 0:
+                        break
+                    if df is None:
+                        df = pd.DataFrame(self.process_ndjson(rows))
+                    else:
+                        df = df.append(pd.DataFrame(self.process_ndjson(rows)))
+                    print("count : ", count)
+                    count += 1
```
Comment (dimberman) [inline src/astro/sql/operators/agnostic_load_file.py:150]: Why do you have this? Is this leftover from debug?

File: src/astro/sql/operators/agnostic_load_file.py [hunk 4]
```diff
@@ -132,16 +135,66 @@ def _load_dataframe(self, path):
...
-            return deserialiser[file_type](
-                stream, **deserialiser_params.get(file_type, {})
-            )
+            if file_type == "ndjson":
+                df = None
+                count = 0
+                while True:
+                    rows = stream.readlines(DEFAULT_CHUNK_SIZE)
+                    if len(rows) == 0:
+                        break
```
Comment (dimberman) [inline src/astro/sql/operators/agnostic_load_file.py:144]: This feels off. Maybe you can do something like run `rows = stream.readlines(DEFAULT_CHUNK_SIZE)` before the loop and then say `while len(rows) != 0`

File: src/astro/sql/operators/agnostic_load_file.py (hunk 1 above)
Comment (dimberman) [inline src/astro/sql/operators/agnostic_load_file.py:162]: why does len(keys) need to be 8? Please add a comment.

File: src/astro/sql/operators/agnostic_load_file.py [hunk 5]
```diff
@@ -132,16 +135,66 @@ def _load_dataframe(self, path):
...
+                if not results.get(k):
+                    results[k] = []
+
+                if k in self.flatten_ndjson:
+                    v = self.flatten_ndjson[k]
+
+                    if isinstance(v, str):
+                        results[k].append(self.ndjson_get_values(v, row))
+                    elif self.isLambda(v):
+                        results[k].append(v(row))
```
Comment (dimberman) [inline src/astro/sql/operators/agnostic_load_file.py:175]: what does isLambda mean?

File: tests/operators/test_agnostic_load_file.py [hunk 6]
```diff
@@ -469,6 +469,80 @@ def test_aql_gcs_file_to_postgres(self):
...
+                    conn_id="postgres_conn",
+                    schema="public",
+                ),
+                "flatten_ndjson": {
+                    "repository": "repository.url",
+                    "actor_attributes": "actor_attributes.company",
+                    "payload": "payload.size",
+                },
+            },
+        )
```
Comment (dimberman) [inline tests/operators/test_agnostic_load_file.py:501]: Let's move away from `create_and_run_task`. It's constricting as we can't have multiple tasks that depend on eachother. Just make the dag and then use `test_utils.run_dag(self.dag)`

File: tests/operators/test_agnostic_load_file.py [hunk 7]
```diff
@@ -469,6 +469,80 @@ def test_aql_gcs_file_to_postgres(self):
...
+                    conn_id="postgres_conn",
+                    schema="public",
+                ),
+                "flatten_ndjson": {
+                    "repository": lambda x: x["repository"].get("url"),
+                    "actor_attributes": lambda x: x["actor_attributes"].get("company"),
+                    "payload": lambda x: x["payload"].get("size"),
+                },
+            },
+        )
```
Comment (dimberman) [inline tests/operators/test_agnostic_load_file.py:538]: same here

**Review by dimberman — CHANGES_REQUESTED** [review]

**Review by tatiana — COMMENTED** [review]
@utkarsharma2 have you considered doing the flattening implicitly, without any changes to the interface of `load_file` (e.g. without adding `flatten_ndjson`)?

Since we are already using pandas, did you consider using `pandas.json_normalize` (https://pandas.pydata.org/docs/reference/api/pandas.json_normalize.html) or some other similar tool?

**Comment by utkarsharma2** [comment]
@tatiana, I like the idea of using pandas json_normalize, it has much richer interface. Maybe we can have a param like 'normalize' which we can pass to json_normalize.

 `json_normalize(data, **normalize)`

File: src/astro/sql/operators/agnostic_load_file.py (hunk 5 above)
Comment (utkarsharma2) [reply src/astro/sql/operators/agnostic_load_file.py]: Have removed this from code, but the idea was to give access to entire row within a lambda function. So user can do operation like creating new col by function execution.

File: tests/operators/test_agnostic_load_file.py (hunk 6 above)
Comment (utkarsharma2) [reply tests/operators/test_agnostic_load_file.py]: Fixed.

File: tests/operators/test_agnostic_load_file.py (hunk 7 above)
Comment (utkarsharma2) [reply tests/operators/test_agnostic_load_file.py]: Fixed.

**Comment by utkarsharma2** [comment]
@dimberman @tatiana this condition can occur for json file type as well, do we need to handle it?

**Review by dimberman — APPROVED** [review]

**Comment by kaxil** [comment]
Can you fix merge conflicts please @utkarsharma2

**Pull request closed without merging**